<?xml version="1.0"?>
<robot name="ghost1" xmlns:xacro="http://ros.org/wiki/xacro">

    <xacro:include filename="$(find ghost_ros)/urdf/swerve_macros.xacro"/>

    <xacro:property name="l_wheel" value="0.2286"/> <!-- 9" center to wheel modules-->
    <xacro:property name="r_wheel" value="0.034925"/> <!-- 1.375" wheel radius-->
    <xacro:property name="width_wheel" value="0.0254"/> <!-- 1" wheel width-->
    <xacro:property name="com_radius" value="0.02"/>
    <xacro:property name="d2r" value="${3.14159/180}"/>

    <link name="base_link">
        <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <sphere radius="${com_radius}"/>
        </geometry>
        <material name="white"/>
        </visual>

        <!--Inertia estimated for solid cylinder of radius 9" and height 3" with mass 7kg-->
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="10"/>
            <inertia
                ixx="0.094839" ixy="0.0" ixz="0.0"
                iyy="0.094839" iyz="0.0"
                izz="0.182903"/>
        </inertial>

        <gazebo>
            <maxVel>3.0</maxVel>
        </gazebo>
    </link>

    <link name="base_footprint">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.00001"/>
            <inertia
                ixx="0.00001" ixy="0.0" ixz="0.0"
                iyy="0.00001" iyz="0.0"
                izz="0.00001"/>
        </inertial>
    </link>    

    <joint name="base_footprint_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin xyz="0 0 ${- r_wheel}" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <!-- Steering Modules -->
    <xacro:steering_macro id="1"></xacro:steering_macro>
    <xacro:steering_macro id="2"></xacro:steering_macro>
    <xacro:steering_macro id="3"></xacro:steering_macro>

    <!-- Wheel Modules -->
    <xacro:wheel_macro id="1" r_wheel="${r_wheel}" width_wheel="${width_wheel}" collision_factor="8"></xacro:wheel_macro>
    <xacro:wheel_macro id="2" r_wheel="${r_wheel}" width_wheel="${width_wheel}" collision_factor="8"></xacro:wheel_macro>
    <xacro:wheel_macro id="3" r_wheel="${r_wheel}" width_wheel="${width_wheel}" collision_factor="8"></xacro:wheel_macro>

    <!-- Joints -->
    <joint name="steering_1" type="continuous">
        <parent link="base_link"/>
        <child link="mod_1"/>
        <origin xyz="${l_wheel/2} ${-l_wheel*0.866} 0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="steering_2" type="continuous">
        <parent link="base_link"/>
        <child link="mod_2"/>
        <origin xyz="${l_wheel/2} ${l_wheel*0.866} 0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="steering_3" type="continuous">
        <parent link="base_link"/>
        <child link="mod_3"/>
        <origin xyz="${-l_wheel} 0 0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="driveshaft_1" type="continuous">
        <parent link="mod_1"/>
        <child link="wheel_1"/>
        <origin xyz="0 0 0" rpy="${90*d2r} 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="driveshaft_2" type="continuous">
        <parent link="mod_2"/>
        <child link="wheel_2"/>
        <origin xyz="0 0 0" rpy="${90*d2r} 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="driveshaft_3" type="continuous">
        <parent link="mod_3"/>
        <child link="wheel_3"/>
        <origin xyz="0 0 0" rpy="${90*d2r} 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <!-- Sensors -->
    <link name="lidar_link">
        <inertial>
            <origin xyz="0.0 0 0.0" rpy="0 0 0"/>
            <mass value="0.125"/>
            <inertia ixx="0.001"  ixy="0"  ixz="0" iyy="0.001" iyz="0" izz="0.001" />
        </inertial>

        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
            <cylinder radius="${0.06985/2}" length="0.0254"/>
            </geometry>
        </visual>
    </link>

    <joint name="lidar_joint" type="fixed">
        <parent link="base_link"/>
        <child link="lidar_link"/>
        <origin xyz="0 0 0.1" rpy="0 0 0"/>
    </joint>

    <link name="camera_base_link">
        <inertial>
            <origin xyz="0.0 0 0.0" rpy="0 0 0"/>
            <mass value="0.125"/>
            <inertia ixx="0.001"  ixy="0.001"  ixz="0.001" iyy="0.001" iyz="0.001" izz="0.001" />
        </inertial>
    </link>

    <joint name="camera_base_joint" type="fixed">
        <parent link="base_link"/>
        <child link="camera_base_link"/>
        <origin xyz="${l_wheel/2} 0 0.33" rpy="0 0 0"/>
    </joint>

    <link name="camera_view_link">
        <inertial>
            <origin xyz="0.0 0 0.0" rpy="0 0 0"/>
            <mass value="0.125"/>
            <inertia ixx="0.001"  ixy="0.001"  ixz="0.001" iyy="0.001" iyz="0.001" izz="0.001" />
        </inertial>
    </link>

    <joint name="camera_view_joint" type="fixed">
        <parent link="camera_base_link"/>
        <child link="camera_view_link"/>
        <origin xyz="0 0 0" rpy="0 0.7853975 0"/>
    </joint>
</robot>